Sources/MongoSwift/EventLoopBoundMongoClient.swift

```diff
@@ -260,4 +260,64 @@ public struct EventLoopBoundMongoClient {
             session: session
         )
     }
+
+    /**
+     * Starts a new `ClientSession` with the provided options. The `ClientSession` will be bound to the `EventLoop`
+     * specified on this `EventLoopBoundMongoClient`. When you are done using this session, you must call
+     * `ClientSession.end()` on it.
+     *
+     * - Parameters:
+     *   - options: The options to use when starting this session.
+     *
+     * - Returns:
+     *   A `ClientSession` that is bound to this `EventLoopBoundMongoClient`'s `EventLoop`.
+     */
+    public func startSession(options: ClientSessionOptions? = nil) -> ClientSession {
+        ClientSession(client: self.client, eventLoop: self.eventLoop, options: options)
+    }
+
+    /**
+     * Starts a new `ClientSession` with the provided options and passes it to the provided closure. The returned
+     * future will be on the `EventLoop` specified on this `EventLoopBoundMongoClient`.
+     * The session is only valid within the body of the closure and will be ended after the body completes.
+     *
+     * - Parameters:
+     *   - options: Options to use when creating the session.
+     *   - sessionBody: A closure which takes in a `ClientSession` and returns an `EventLoopFuture<T>`.
+     *
+     * - Returns:
+     *    An `EventLoopFuture<T>`, the return value of the user-provided closure, on this `EventLoopBoundMongoClient`'s
+     *    `EventLoop`.
+     *
+     *    If the future fails, the error is likely one of the following:
+     *    - `CompatibilityError` if the deployment does not support sessions.
+     *    - `MongoError.LogicError` if this client has already been closed.
+     */
+    public func withSession<T>(
+        options: ClientSessionOptions? = nil,
+        _ sessionBody: (ClientSession) throws -> EventLoopFuture<T>
+    ) -> EventLoopFuture<T> {
+        let promise = self.client.operationExecutor.makePromise(of: T.self, on: self.eventLoop)
+        let session = self.startSession(options: options)
+        do {
+            let bodyFuture = try sessionBody(session)
```

Comment:
@patrickfreed  What would be an example of when this could happen or go wrong? I am trying to trigger a fail for the `withSession` test case in order to ensure this works. The `withSession` returns a promise that's on bound event loop even when you provide a future that is not on the same event loop as `EventLoopBoundMongoClient`.
Before making any changes to withSessions I am trying to make this fail:
```
let withSessionFuture = elBoundClient.withSession { _ in
   // do something and return a future that is on a different event loop
 }
// test withSessionFuture.eventLoop equals expectedEventLoop
```
but it passes

Reply:
The event loop of the final returned future will always be correct because we set it to the correct one when we create the promise that we use to return the future. However, the stuff that happens in `ClientSession.end` outside of the executor will get executed on whatever event loop the user used for the user's future because that stuff is happening as part of a callback on that future. This means the `switch` on `self.state` and the assignment of `self.state` to `.ended` for example will be happening on the wrong event loop. You can test this by adding a `self.eventLoop.assertInEventLoop()` to the top of `ClientSession.end` (don't leave this here, because `end` could be called from outside of the event loop too).

This model of concurrency safety is a little tricky and has taken me a while to wrap my head around. The idea is this: to ensure exclusive access to certain state, only ever interact with it while on a specific event loop, since an event loop is only ever executing a single thing at a given time. It can be _waiting_ for events from many things at a time though. When it gets an event from a thing (i.e. a future produces a result), it processes that event and then invokes any associated callback, but it only does so one at a time (you can think of a single event loop as a single thread: a single thread can't read from multiple sockets at once but it can `select` over them and then read the ready ones).

For instance, consider the following:
```swift
func asyncWork(el: EventLoop) -> EventLoopFuture<Int> { ... }

func main() {
    let elg = MultiThreadedEventLoopGroup(numberOfThreads: 4)

    let el1 = elg.next()
    let el2 = elg.next()

    var sharedState = 1

    let fut1 = asyncWork(el: el1).map { sharedState = sharedState + 1 }
    let fut2 = asyncWork(el: el2).map { sharedState = sharedState + 1 }

    try fut1.and(fut2).wait()

    print(sharedState)
}
```
In this example, we access `sharedState` from two separate event loops simultaneously. This means that the callbacks will race and they could potentially interfere with each other. Consider the case where `fut1`'s callback increments `sharedState` but only does so after `fut2`'s callback started. We could see the following sequence of events:

```
fut1 callback starts
fut2 callback starts
fut1 reads sharedState = 1
fut2 reads sharedState = 1
fut 1 sets sharedState to 1 + 1 = 2
fut2 sets sharedState to 1 + 1 = 2
print(2)
```
In other cases this will print 3 as expected.

Now if we drop `el2` and create two futures both on `el1`, then only one of the two callbacks will be executed at a time (the one on whichever future the event loop gets an event from first). This means it will _always_ print 3 as expected.

```
fut2 callback starts
fut2 reads sharedState = 1
fut2 sets sharedState to 1 + 1 = 2
fut1 callback starts
fut1 reads sharedState = 2
fut1 sets sharedState to 2 + 1 = 3
print(3)
```

Note that this order could be flipped and `fut1`s callback could come first since the event loop is waiting for events from either of them, it just depends which future produces a result first. 

Getting back to the current situation, I'm not sure if there's really an easy way to test this unfortunately, since as you noted the final event loops are always correct. I think the test ensuring the event loops on the returned futures are correct should be sufficient then.

Note: the terminology I used in this may not be the most accurate. I'm still trying to understand this stuff well myself.

Reply:
@patrickfreed Thank you! This was very helpful! 😊